pkg/config/mesh/meshwatcher/mesh.go

```diff
@@ -0,0 +1,112 @@
+	"google.golang.org/protobuf/proto"
+
+	meshconfig "istio.io/api/mesh/v1alpha1"
+	"istio.io/istio/pkg/config/mesh"
+	"istio.io/istio/pkg/kube/krt"
+	"istio.io/istio/pkg/util/protomarshal"
+)
+
+// WatcherCollection is an interface to describe an object that implements both the legacy mesh.Watcher interface and the
+// new krt interface.
+type WatcherCollection interface {
+	mesh.Watcher
+	krt.Singleton[MeshConfigResource]
+}
+
+// ConfigAdapter wraps a MeshConfig collection into a mesh.Watcher interface.
+func ConfigAdapter(configuration krt.Singleton[MeshConfigResource]) WatcherCollection {
+	return adapter{configuration}
+}
+
+// adapter is a helper to expose the krt collection using the older mesh.Watcher interface
+type adapter struct {
+	krt.Singleton[MeshConfigResource]
+}
+
+var _ mesh.Watcher = adapter{}
+
+// Mesh returns the current MeshConfig
+func (a adapter) Mesh() *meshconfig.MeshConfig {
+	// Just get the value; we know there is always one set due to the way the collection is setup.
+	v := a.Singleton.Get()
+	return v.MeshConfig
+}
+
+// AddMeshHandler registers a callback that will be called anytime the MeshConfig changes.
+// Usually a handler would then call .Mesh() to get the new state.
+// The returned WatcherHandlerRegistration can be used to un-register the handler at a later time.
+func (a adapter) AddMeshHandler(h func()) *mesh.WatcherHandlerRegistration {
+	active := uatomic.NewBool(true)
+	// The mesh.Watcher allows unregistering, while krt currently doesn't
+	// When we remove, mark this as "not active" so we skip events.
+	reg := mesh.NewWatcherHandlerRegistration(func() {
+		active.Store(false)
+	})
+	// Do not run initial state to match existing semantics
+	a.Singleton.AsCollection().RegisterBatch(func(o []krt.Event[MeshConfigResource], initialSync bool) {
+		if active.Load() {
+			h()
+		}
+	}, false)
+	return reg
+}
+
+// DeleteMeshHandler removes a previously registered handler.
+func (a adapter) DeleteMeshHandler(registration *mesh.WatcherHandlerRegistration) {
+	registration.Remove()
+}
+
+// MeshConfigResource holds the current MeshConfig state
+type MeshConfigResource struct {
+	*meshconfig.MeshConfig
+}
+
+func (m MeshConfigResource) ResourceName() string { return "MeshConfigResource" }
+
+func (m MeshConfigResource) Equals(other MeshConfigResource) bool {
+	return proto.Equal(m.MeshConfig, other.MeshConfig)
+}
+
+// MeshNetworksResource holds the current MeshNetworks state
+type MeshNetworksResource struct {
+	*meshconfig.MeshNetworks
+}
+
+func (m MeshNetworksResource) ResourceName() string { return "MeshNetworksResource" }
+
+func (m MeshNetworksResource) Equals(other MeshNetworksResource) bool {
+	return proto.Equal(m.MeshNetworks, other.MeshNetworks)
+}
+
+// NetworksAdapter wraps a MeshNetworks collection into a mesh.NetworksWatcher interface.
+func NetworksAdapter(configuration krt.Singleton[MeshNetworksResource]) mesh.NetworksWatcher {
+	return networksAdapter{configuration}
+}
```

Comment:
Do these networks-related bits make more sense in `pkg/config/mesh/meshwatcher/networks.go` instead?

Same goes for  `func PrettyFormatOfMeshNetworks...`